[ ResultsDB ] Modify 'Find Regression Point' button behavior on Results Database

iambryyy opened the pull request:
#### 8346f784b526e4607316f05ff4831f204b64a972
<pre>
[ ResultsDB ] Modify &apos;Find Regression Point&apos; button behavior on Results Database
<a href="https://rdar.apple.com/159977488">rdar://159977488</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=298484">https://bugs.webkit.org/show_bug.cgi?id=298484</a>

Reviewed by Jonathan Bedard.

Included a &apos;regression range&apos; that will assist in narrowing regression ranges
by highlighting the earliest and latest non-successful run within an analysis
window.

* Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
(TimelineFromEndpoint.prototype.this.toggleShowRegressionHighlight): Logic for range highlight button
(TimelineFromEndpoint.prototype.detectRegression): Include range for flaky
(TimelineFromEndpoint.prototype.getRangeUuids): Logic for declaring where to highlight range
(TimelineFromEndpoint.prototype.highlightRange): Visualizes the regression range

Canonical link: <a href="https://commits.webkit.org/303058@main">https://commits.webkit.org/303058@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ccf24afa98f41de169433deedd4ab730d77ab150

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131077 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3404 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42038 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138518 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/82769 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132947 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3394 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3245 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99857 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/82769 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134023 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2424 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117342 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80562 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/130425 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2346 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/35463 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81765 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110942 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35978 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141012 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3147 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/2814 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108374 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/130505 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3195 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113672 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108327 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2378 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32094 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56205 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20403 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3214 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3034 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3235 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3144 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -17,7 +17,7 @@
 // SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 // INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
-// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)`
```
What's going on here?

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -761,7 +795,75 @@ class TimelineFromEndpoint {
         if (!found)
             this.searchScale(null);
     };
+    highlightRange(baseDotRenderer, seriesResults, uuidRange) {
```
I think we want to separate the drawing of the range from the improved algorithm. Especially since it is my understanding that the drawing of the range revealed deficiencies in the algorithm.

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -761,7 +795,75 @@ class TimelineFromEndpoint {
+            const range = this.regRange;
+            if (!range) return baseDotRenderer(data, context, x, y);
+
+            if (!uuidRange.has(range.leftUuid) || !uuidRange.has(range.rightUuid)) {
```
What is the relationship between `uuidRange` and `range`? I would expect this to be `if (!uuidRange || !uuidRange.hash...`

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -761,7 +795,75 @@ class TimelineFromEndpoint {
+                return baseDotRenderer(data, context, x, y);
+            }
+
+            const rangeKey = range.leftUuid + '→' + range.rightUuid;
```
I don't understand why this is needed.

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -761,7 +795,75 @@ class TimelineFromEndpoint {
+            const li = idxByUuid.get(range.leftUuid);
+            const ri = idxByUuid.get(range.rightUuid);
+            if (li == null || ri == null) return baseDotRenderer(data, context, x, y);
+
```
Something is wrong with the spacing here

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -761,7 +795,75 @@ class TimelineFromEndpoint {
+                    return !!failureType && failureType !== 'success';
+                });
+                if (!anyFailInRange) {
+                    return baseDotRenderer(data, context, x, y);
```
Why would we want to return if there were no failures in the range? We want to see if our algorithm reports a range as being to blame but does not contain any failures, right?

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -770,26 +872,18 @@ class TimelineFromEndpoint {
-                left: 0,
-            },
+            state: {show: false, top: 0, left: 0},
             onElementMount: (element) => {
```
Unnecessary format churn

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -312,8 +313,8 @@ function statsForSingleResult(result) {
         const idForType = Expectations.stringToStateId(Expectations.failureTypeMap[type]);
-        stats[`tests_${type}`] = actualId > idForType  ? 0 : 1;
-        stats[`tests_unexpected_${type}`] = unexpectedId > idForType  ? 0 : 1;
+        stats[`tests_${type}`] = actualId > idForType ? 0 : 1;
```
For these, you can use the unary operator to cast the booleans into integers:
```
stats[`tests_${type}`] = +(actualId <= idForType)
stats[`tests_unexpected_${type}`] = +(unexpectedId <= idForType)
```

More: https://stackoverflow.com/a/7820695

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -662,6 +672,32 @@ class TimelineFromEndpoint {
+
+        for (let j = start; j <= end; j++) {
+            const { failureType } = this.getTestResultStatus(allConfigResults[j], willFilterExpected);
+            const failure = !!failureType && failureType !== 'success';
```
From offline discussion: you have a few places you do this where you should double-check to make sure that `failureType` will ever be falsy. Otherwise, the first condition of `!!failureType` will never be `false` and, thus, is unnecessary.

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -761,7 +795,75 @@ class TimelineFromEndpoint {
+            }
+
+            const range = this.regRange;
+            if (!range) return baseDotRenderer(data, context, x, y);
```
nit: you're using a lot of different conventions for one-line returns under `if` conditions. Here you have a one-liner, but on L806 you have it within `{}`, and on L751 you have separate lines without `{}`. I'd recommend choosing one for consistency.

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -770,26 +872,18 @@ class TimelineFromEndpoint {
-                        element.style.display = 'none';
-                }
+                if ('show' in stateDiff)
+                    element.style.display = stateDiff.show ? 'block' : 'none';
```
Genuinely not sure if this needs to be nested, are you sure this doesn't work?
```
element.style.display = 'show' in stateDiff && stateDiff.show ? 'block' : 'none';
```
You do this in a few places, so maybe it's right? But worth double-checking.

bls1999 commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -1457,30 +1569,30 @@ function Legend(callback=null, plural=false, defaultWillFilterExpected=false, fl
-                                return;
-                            ToolTip.setByElement(
-                                `<div class="content">
+        const dot = REF.createRef({
```
You should get rid of this formatting churn.

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -40,6 +40,7 @@ let willFilterExpected = false;
 let showTestTimes = false;
 let showTestFlakiness = false;
 let showNumberOfFlakyTests = false;
+let showRange = false;
```
Where do we use this?

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -312,8 +313,8 @@ function statsForSingleResult(result) {
-        stats[`tests_${type}`] = actualId > idForType  ? 0 : 1;
-        stats[`tests_unexpected_${type}`] = unexpectedId > idForType  ? 0 : 1;
+            stats[`tests_${type}`] = +(actualId <= idForType)
+            stats[`tests_unexpected_${type}`] = +(unexpectedId <= idForType)
```
Doesn't seem like the churn here is valuable.

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -839,7 +943,7 @@ class TimelineFromEndpoint {
 
                 let tag = [];
-                let pushTag = (value, prefix='', suffix='') => {
+                const pushTag = (value, prefix='', suffix='') => {
```
Why the churn?

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -874,7 +978,7 @@ class TimelineFromEndpoint {
                     pushTag(time);
                 if (showTestFlakiness && data.flakiness_num_passes && data.flakiness_num_tries) {
-                    let flakiness = 1 - (data.flakiness_num_passes / data.flakiness_num_tries);
+                    const flakiness = 1 - (data.flakiness_num_passes / data.flakiness_num_tries);
```
Why the churn?

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -1019,7 +1123,7 @@ class TimelineFromEndpoint {
             // Create a list of configurations to display with SDKs stripped
             let mappedChildrenConfigs = {};
-            let childrenConfigsBySDK = {}
+            let childrenConfigsBySDK = {};
```
Churn, yes, but keep this one.

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -1278,15 +1394,18 @@ class TimelineFromEndpoint {
         const nextRegressButtonRef = REF.createRef({
             state: {
-                disabled: true,
+                disabled: true
```
Why the churn here?

JonWBedard commented on Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
```diff
@@ -1303,7 +1422,7 @@ class TimelineFromEndpoint {
         const previousRegressButtonRef = REF.createRef({
             state: {
-                disabled: true,
+                disabled: true
```
Why the churn here?

iambryyy committed:
[ ResultsDB ] Modify 'Find Regression Point' button behavior on Results Database
rdar://159977488
https://bugs.webkit.org/show_bug.cgi?id=298484

Reviewed by Jonathan Bedard.

Included a 'regression range' that will assist in narrowing regression ranges
by highlighting the earliest and latest non-successful run within an analysis
window.

* Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:
(TimelineFromEndpoint.prototype.this.toggleShowRegressionHighlight): Logic for range highlight button
(TimelineFromEndpoint.prototype.detectRegression): Include range for flaky
(TimelineFromEndpoint.prototype.getRangeUuids): Logic for declaring where to highlight range
(TimelineFromEndpoint.prototype.highlightRange): Visualizes the regression range

Canonical link: https://commits.webkit.org/303058@main
